Python/marshal.c

```diff
@@ -509,17 +509,27 @@ w_complex_object(PyObject *v, char flag, WFILE *p)
             return;
         }
         W_SIZE(n, p);
-        it = PyObject_GetIter(v);
-        if (it == NULL) {
+        l = PySequence_List(v);
+        if (l == NULL) {
             p->depth--;
             p->error = WFERR_UNMARSHALLABLE;
             return;
         }
-        while ((value = PyIter_Next(it)) != NULL) {
-            w_object(value, p);
-            Py_DECREF(value);
+        for (i = 0; i < n; i++) {
+            value = PyList_GetItem(l, i);
```

Comment:
I think the `PyList_GetItem()` and `PyList_SetItem()` calls added in this patch should be replaced with the `PyList_GET_ITEM()` and `PyList_SET_ITEM()` macros.

Also, `PyMarshal_WriteObjectToString()` should be checked for failure.